src/ray/common/asio/dispatchable.h

```diff
@@ -0,0 +1,64 @@
+#pragma once
+#include <functional>
+#include <utility>
+
+#include "ray/common/asio/instrumented_io_context.h"
+
+namespace ray {
+
+// Wrapper for a std::function that includes an instrumented_io_context for dispatching.
+// On `Dispatch`, the function is called with the provided arguments, dispatched onto the
+// provided io_context.
+//
+// The io_context must outlive the Dispatchable object.
+template <typename FuncType>
+class Dispatchable {
+ public:
+  Dispatchable(std::function<FuncType> func, instrumented_io_context &io_context)
+      : func_(std::move(func)), io_context_(&io_context) {}
+
+  explicit Dispatchable(nullptr_t) : func_(nullptr), io_context_(nullptr) {}
+
+  template <typename... Args>
+  void DispatchIfNonNull(const std::string &name, Args &&...args) const {
+    if (func_) {
+      RAY_CHECK(io_context_ != nullptr);
+      io_context_->dispatch(
+          [func = func_,
+           args_tuple = std::make_tuple(std::forward<Args>(args)...)]() mutable {
+            std::apply(func, std::move(args_tuple));
+          },
+          name);
+    }
+  }
+
+  std::function<FuncType> AsDispatchedFunction(const std::string &name) && {
```

Comment:
This should be the only public API (i.e. DispatchIfNotNull should be private)?

Also I'm wondering whether we need the Dispatchable class. Don't we only need a static function 


```
template <class FuncType>
std::function<FuncType> AsDispatchedFunction(std::function<FuncType> func, io_context, name) {
  return [=](auto &&...args) {
    io_context.dispatch([] {func(args)} );
  }
} 
```

Reply:
I want `Dispatchable` to be put in InternalKVInterface params so it *ensures* the callback is dispatched